### DISCUSSION TITLE
"Virtual" AC device created via a Smart IR

### DISCUSSION THREAD

**Discussion by lageorgem** [make-all/tuya-local#933] [Q&A] [answered]
Hello,

I've been looking through the issues here, and I did find some requests for devices similar to mine (#929 has the same category and #652 has the same mapping).

I want to ask if it is possible to add the AC devices controlled by my smart IR via tuya-local. 
I have tried adding the IR device, but it only shows up with temperature and humidity, while adding my AC's device ID as sub device ID doesn't work at all - it times out, as opposed to giving the "not supported" message.

Has anyone had any luck in the past with this? Current workaround is a HA custom switch that calls google assistant's API.

**Comment by make-all** [comment, marked as answer]
Because IR sub devices are one way, they do not return any data to identify them by.  Currently this prevents them from being identified, as the current device identification method requires some data from the device to match against.

**Comment by lageorgem** [comment]
At least in my case, it seems like the `/v2.0/cloud/thing/{device_id}/shadow/properties` API returns some data, including dp_ids:

```json
{
  "result": {
    "properties": [
      {
        "code": "control",
        "custom_name": "",
        "dp_id": 1,
        "time": 1674561759838,
        "value": "send_ir"
      },
      {
        "code": "study_code",
        "custom_name": "",
        "dp_id": 2,
        "time": 1674561759838
      },
      {
        "code": "ir_code",
        "custom_name": "",
        "dp_id": 3,
        "time": 1674561759838,
        "value": ""
      },
      {
        "code": "key_code",
        "custom_name": "",
        "dp_id": 4,
        "time": 1674561759838,
        "value": ""
      },
      {
        "code": "key_code2",
        "custom_name": "",
        "dp_id": 5,
        "time": 1674561759838,
        "value": ""
      },
      {
        "code": "key_code3",
        "custom_name": "",
        "dp_id": 6,
        "time": 1674561759838,
        "value": ""
      },
      {
        "code": "key_study",
        "custom_name": "",
        "dp_id": 7,
        "time": 1674561759838
      },
      {
        "code": "key_study2",
        "custom_name": "",
        "dp_id": 8,
        "time": 1674561759838
      },
      {
        "code": "key_study3",
        "custom_name": "",
        "dp_id": 9,
        "time": 1674561759838
      },
      {
        "code": "delay_time",
        "custom_name": "",
        "dp_id": 10,
        "time": 1674561759838,
        "value": 0
      },
      {
        "code": "key_code4",
        "custom_name": "",
        "dp_id": 11,
        "time": 1674561759838,
        "value": ""
      },
      {
        "code": "key_study4",
        "custom_name": "",
        "dp_id": 12,
        "time": 1674561759838
      },
      {
        "code": "type",
        "custom_name": "",
        "dp_id": 13,
        "time": 1674561759838,
        "value": 0
      },
      {
        "code": "switch_power",
        "custom_name": "",
        "dp_id": 101,
        "time": 1690404360694,
        "value": false
      },
      {
        "code": "mode",
        "custom_name": "",
        "dp_id": 102,
        "time": 1690404277957,
        "value": "0"
      },
      {
        "code": "temperature",
        "custom_name": "",
        "dp_id": 103,
        "time": 1690404277957,
        "value": 20
      },
      {
        "code": "fan",
        "custom_name": "",
        "dp_id": 104,
        "time": 1690404277957,
        "value": "0"
      },
      {
        "code": "swing",
        "custom_name": "",
        "dp_id": 105,
        "time": 1674561759838,
        "value": false
      },
      {
        "code": "ir_send",
        "custom_name": "",
        "dp_id": 201,
        "time": 1674561759838,
        "value": ""
      },
      {
        "code": "ir_study_code",
        "custom_name": "",
        "dp_id": 202,
        "time": 1674561759838
      }
    ]
  },
  "success": true,
  "t": 1690531316304,
  "tid": "0563ee1b2d1d11ee94e0a2e0adfc0b14"
}
```

Wouldn't this be enough to create a device? And it's not even that the device is unidentified, I simply get a timeout when trying to connect the IR device and add the sub device ID. Am I doing it correctly?

**Reply by make-all** [reply]
That is cloud a cloud API function, and those are commands to the device. Normally there should be a "status" section as well, with the values that come from the device.  The timeout is occurring because no status comes from the device.

**Reply by lageorgem** [reply]
I understand, I also did a bit of research and seems none of these IR subdevices report anything. Would it be viable to implement a feature that allows such devices to be manually paired with a configuration file? I would presume these devices can be controlled just fine, but the config matching code is unable to work with it.

**Reply by make-all** [reply]
I have one of these IR devices myself that I bought thinking it would help to debug the sub-device work, so it is on my radar and something I hope to solve one day.  But I think it is best to implement real network discovery first and see if that helps before going the manual route, which I think many users would be misled into trying for devices that truly aren't supported.  Maybe if the configs for undetectable devices were separate from others it might be possible to avoid this to some degree.
